File: src/main/栈/t85/MaximalRectangle.java
package 栈.t85;

import java.util.LinkedList;

/**
 * @auther: sotowang
 * @date: 2019/12/31 22:14
 */
public class MaximalRectangle {
    //与84题类似
    public int maximalRectangle(char[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0] == null || matrix[0].length == 0) {
            return 0;
        }
        int rows = matrix.length;
        int cols = matrix[0].length;
        int[] rec = new int[cols];
        int max = 0;
        //将最大连续1做为矩形的高度
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (matrix[i][j] == '0') {
                    rec[j] = 0;
                } else {
                    rec[j]++;
                }
            }
            max = Math.max(calculateLargestArea(rec), max);
        }
        return max;
    }

    private int calculateLargestArea(int[] matrix) {
        LinkedList<Integer> stack = new LinkedList<>();
        stack.push(0);
        int max = 0;
        for (int i = 1; i < matrix.length; i++) {
            while (!stack.isEmpty() && matrix[i] < matrix[stack.peek()]) {
                int popIndex = stack.pop();
                int topIndex = stack.peek() == null ? -1 : stack.peek();
                max = Math.max((i - topIndex - 1) * matrix[popIndex], max);
            }
            stack.push(i);
        }
        while (!stack.isEmpty()) {
            int popIndex = stack.pop();
            int topIndex = stack.peek() == null ? -1 : stack.peek();
            max = Math.max((matrix.length - topIndex - 1) * matrix[popIndex], max);
        }
        return max;
    }

    public static void main(String[] args) {
        //{"1","0","1","0","0"}
        //{"1","0","1","1","1"}
        //{"1","1","1","1","1"}
        //{"1","0","0","1","0"}
        char[][] matrix = new char[][]{
                {'1', '0', '1', '0', '0'},
                {'1', '0', '1', '1', '1'},
                {'1', '1', '1', '1', '1'},
                {'1', '0', '0', '1', '0'}
        };
        assert new MaximalRectangle().maximalRectangle(matrix) == 6;
        char[][] matrix1 = new char[][]{
                {'1', '0'},
                {'0', '1'},
        };
        assert new MaximalRectangle().maximalRectangle(matrix1) == 1;



    }
}

File: src/main/栈/最大矩形/t85/maximalRectangle.go
package main

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
func maximalRectangle(matrix [][]byte) int {
	if matrix == nil || len(matrix) == 0 {
		return 0
	}
	rows, cols := len(matrix), len(matrix[0])
	rec := make([]int, cols)
	maxVal := 0
	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			if matrix[i][j] == '0' {
				rec[j] = 0
			} else {
				rec[j]++
			}
		}
		maxVal = max(calculateLargestArea(rec), maxVal)
	}
	return maxVal
}
func calculateLargestArea(matrix []int) int {
	stack := []int{0}
	maxVal, n := 0, len(matrix)
	for i := 1; i < n; i++ {
		for len(stack) > 0 && matrix[i] < matrix[stack[len(stack)-1]] {
			popIndex, topIndex := stack[len(stack)-1], -1
			stack = stack[:len(stack)-1]
			if len(stack) != 0 {
				topIndex = stack[len(stack)-1]
			}
			maxVal = max((i-topIndex-1)*matrix[popIndex], maxVal)
		}
		stack = append(stack, i)

	}
	for len(stack) > 0 {
		popIndex, topIndex := stack[len(stack)-1], -1
		stack = stack[:len(stack)-1]
		if len(stack) != 0 {
			topIndex = stack[len(stack)-1]
		}
		maxVal = max((len(matrix)-topIndex-1)*matrix[popIndex], maxVal)
	}
	return maxVal
}

func main() {
	matrix := [][]byte{{'1', '0', '1', '0', '0'},
		{'1', '0', '1', '1', '1'},
		{'1', '1', '1', '1', '1'},
		{'1', '0', '0', '1', '0'}}
	maximalRectangle(matrix)
}
